TST: TestEig.test_singular failing tolerance with generic BLAS installed

### Describe your issue.

`TestEig.test_singular` from `scipy.linalg.tests.test_decomp` was recently activated in PR #20203 .
It is failing in tests for the Debian build of  scipy 1.13.1 on several architectures including amd64. 
https://buildd.debian.org/status/package.php?p=scipy&suite=experimental
amd64: https://buildd.debian.org/status/fetch.php?pkg=scipy&arch=amd64&ver=1.13.1-1exp1&stamp=1717613707&raw=0
armhf: https://buildd.debian.org/status/fetch.php?pkg=scipy&arch=armhf&ver=1.13.1-1exp1&stamp=1717617698&raw=0
mips64el: https://buildd.debian.org/status/fetch.php?pkg=scipy&arch=mips64el&ver=1.13.1-1exp1&stamp=1717617903&raw=0

The absolute tolerance is just missed, set to `atol=1e-13`, it got `Max absolute difference: 1.53003533e-13`
In principle this was anticipated, the test sets `atol_homog=5e-13`. The problem is that `atol_homog` (and `rtol_homog`) is not being used at https://github.com/scipy/scipy/blob/59d8ebca54a2c14390eacc1823f23fafbd14c021/scipy/linalg/tests/test_decomp.py#L233

It passed on arm64, and on my local amd64 system, which has openblas installed. The BLAS on the failing systems is the generic netlib implementation `libblas-dev`

(debian packages build against generic blas, taking advantage of BLAS ABI compatibility with users expected to install optimised BLAS such as openblas on their systems)

### Reproducing Code Example

```python
# with netlib generic blas installed, not openblas
python3 -c "import scipy.linalg; scipy.linalg.test(extra_argv=['-k', 'test_decomp and TestEig and test_singular', '-v'])"
```


### Error message

```shell
____________________________ TestEig.test_singular _____________________________

self = <scipy.linalg.tests.test_decomp.TestEig object at 0x7fcdfaafba40>

    def test_singular(self):
        # Example taken from
        # https://web.archive.org/web/20040903121217/http://www.cs.umu.se/research/nla/singular_pairs/guptri/matlab.html
        A = array([[22, 34, 31, 31, 17],
                   [45, 45, 42, 19, 29],
                   [39, 47, 49, 26, 34],
                   [27, 31, 26, 21, 15],
                   [38, 44, 44, 24, 30]])
        B = array([[13, 26, 25, 17, 24],
                   [31, 46, 40, 26, 37],
                   [26, 40, 19, 25, 25],
                   [16, 25, 27, 14, 23],
                   [24, 35, 18, 21, 22]])
    
        with np.errstate(all='ignore'):
>           self._check_gen_eig(A, B, atol_homog=5e-13)

A          = array([[22, 34, 31, 31, 17],
       [45, 45, 42, 19, 29],
       [39, 47, 49, 26, 34],
       [27, 31, 26, 21, 15],
       [38, 44, 44, 24, 30]])
B          = array([[13, 26, 25, 17, 24],
       [31, 46, 40, 26, 37],
       [26, 40, 19, 25, 25],
       [16, 25, 27, 14, 23],
       [24, 35, 18, 21, 22]])
self       = <scipy.linalg.tests.test_decomp.TestEig object at 0x7fcdfaafba40>

scipy/linalg/tests/test_decomp.py:262: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
scipy/linalg/tests/test_decomp.py:222: in _check_gen_eig
    assert_allclose(res[:, i], 0,
        A          = array([[22, 34, 31, 31, 17],
       [45, 45, 42, 19, 29],
       [39, 47, 49, 26, 34],
       [27, 31, 26, 21, 15],
       [38, 44, 44, 24, 30]])
        B          = array([[13, 26, 25, 17, 24],
       [31, 46, 40, 26, 37],
       [26, 40, 19, 25, 25],
       [16, 25, 27, 14, 23],
       [24, 35, 18, 21, 22]])
        B0         = array([[13, 26, 25, 17, 24],
       [31, 46, 40, 26, 37],
       [26, 40, 19, 25, 25],
       [16, 25, 27, 14, 23],
       [24, 35, 18, 21, 22]])
        atol_homog = 5e-13
        beta_nonzero = array([ True,  True,  True,  True, False])
        i          = 0
        length     = array([1., 1., 1., 1., 1.])
        msg        = '\narray([[22, 34, 31, 31, 17],\n       [45, 45, 42, 19, 29],\n       [39, 47, 49, 26, 34],\n       [27, 31, 26, 21, 1...      [31, 46, 40, 26, 37],\n       [26, 40, 19, 25, 25],\n       [16, 25, 27, 14, 23],\n       [24, 35, 18, 21, 22]])'
        perm       = array([4, 0, 3, 2, 1])
        permt      = array([4, 0, 3, 2, 1])
        res        = array([[-1.04410300e-14+0.00000000e+00j, -1.42108547e-14+0.00000000e+00j,
         9.02953242e-15+6.61744490e-23j,  9....j,
         1.12513895e-14+0.00000000e+00j,  1.12513895e-14-2.64697796e-23j,
                   -inf           +nanj]])
        rtol_homog = 1e-13
        self       = <scipy.linalg.tests.test_decomp.TestEig object at 0x7fcdfaafba40>
        val1       = array([[ 3.99680289e-14+0.00000000e+00j,  7.55164063e+00+0.00000000e+00j,
         3.55271368e-15+7.31689048e-08j,  3....j,
         3.55271368e-15+1.46337810e-07j,  3.55271368e-15-1.46337810e-07j,
         5.79566522e+00+0.00000000e+00j]])
        val2       = array([[ 5.04090589e-14-0.00000000e+00j,  7.55164063e+00+0.00000000e+00j,
        -5.47681874e-15+7.31689048e-08j, -5....j,
        -7.69867584e-15+1.46337810e-07j, -7.69867584e-15-1.46337810e-07j,
                    inf           +nanj]])
        vr         = array([[-0.40436804+0.00000000e+00j,  0.95130136+0.00000000e+00j,
        -0.08819712-1.55281010e-09j, -0.08819712+1.5...00000000e+00j,
         0.63139317+5.63409302e-09j,  0.63139317-5.63409302e-09j,
        -0.65411867+0.00000000e+00j]])
        w          = array([-9.45925911e+00+0.00000000e+00j,  2.00000000e+00+0.00000000e+00j,
       -6.62886895e-16+1.27101676e-08j, -6.62886895e-16-1.27101676e-08j,
                   inf+0.00000000e+00j])
        wh         = array([-9.45925911e+00+0.00000000e+00j,  2.00000000e+00+0.00000000e+00j,
       -6.62886895e-16+1.27101676e-08j, -6.62886895e-16-1.27101676e-08j])
        wt         = array([-9.45925911e+00+0.00000000e+00j,  2.00000000e+00+0.00000000e+00j,
       -6.62886895e-16+1.27101676e-08j, -6.62886895e-16-1.27101676e-08j,
                   inf+0.00000000e+00j])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<function assert_allclose.<locals>.compare at 0x7fcdbaeef100>, array([-1.04410300e-14+0.j, -1.53003533e-13+0.j, -9.37126904e-14+0.j,
       -1.31779356e-14+0.j, -9.19363335e-14+0.j]), array(0))
kwds = {'equal_nan': True, 'err_msg': '\narray([[22, 34, 31, 31, 17],\n       [45, 45, 42, 19, 29],\n       [39, 47, 49, 26, ..., 14, 23],\n       [24, 35, 18, 21, 22]])', 'header': 'Not equal to tolerance rtol=1e-13, atol=1e-13', 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-13, atol=1e-13
E           
E           array([[22, 34, 31, 31, 17],
E                  [45, 45, 42, 19, 29],
E                  [39, 47, 49, 26, 34],
E                  [27, 31, 26, 21, 15],
E                  [38, 44, 44, 24, 30]])
E           array([[13, 26, 25, 17, 24],
E                  [31, 46, 40, 26, 37],
E                  [26, 40, 19, 25, 25],
E                  [16, 25, 27, 14, 23],
E                  [24, 35, 18, 21, 22]])
E           Mismatched elements: 1 / 5 (20%)
E           Max absolute difference: 1.53003533e-13
E           Max relative difference: inf
E            x: array([-1.044103e-14+0.j, -1.530035e-13+0.j, -9.371269e-14+0.j,
E                  -1.317794e-14+0.j, -9.193633e-14+0.j])
E            y: array(0)

args       = (<function assert_allclose.<locals>.compare at 0x7fcdbaeef100>, array([-1.04410300e-14+0.j, -1.53003533e-13+0.j, -9.37126904e-14+0.j,
       -1.31779356e-14+0.j, -9.19363335e-14+0.j]), array(0))
func       = <function assert_array_compare at 0x7fce00756ca0>
kwds       = {'equal_nan': True, 'err_msg': '\narray([[22, 34, 31, 31, 17],\n       [45, 45, 42, 19, 29],\n       [39, 47, 49, 26, ..., 14, 23],\n       [24, 35, 18, 21, 22]])', 'header': 'Not equal to tolerance rtol=1e-13, atol=1e-13', 'verbose': True}
self       = <contextlib._GeneratorContextManager object at 0x7fce01163560>

/usr/lib/python3.12/contextlib.py:81: AssertionError
```


### SciPy/NumPy/Python version and system information

```shell
$ python3 -c "import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()"
1.13.1 1.26.4 sys.version_info(major=3, minor=11, micro=9, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/include/x86_64-linux-gnu
    lib directory: /usr/lib/x86_64-linux-gnu
    name: blas
    openblas configuration: unknown
    pc file directory: /usr/lib/x86_64-linux-gnu/pkgconfig
    version: 3.12.0
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/include/x86_64-linux-gnu
    lib directory: /usr/lib/x86_64-linux-gnu
    name: lapack
    openblas configuration: unknown
    pc file directory: /usr/lib/x86_64-linux-gnu/pkgconfig
    version: 3.12.0
  pybind11:
    detection method: pkgconfig
    include directory: /include
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    args: -g, -O2, -Werror=implicit-function-declaration, -ffile-prefix-map=/build/reproducible-path/scipy-1.12.0=.,
      -fstack-protector-strong, -fstack-clash-protection, -Wformat, -Werror=format-security,
      -fcf-protection, -Wdate-time, -D_FORTIFY_SOURCE=2
    commands: cc
    linker: ld.bfd
    linker args: -Wl,-z,relro, -g, -O2, -Werror=implicit-function-declaration, -ffile-prefix-map=/build/reproducible-path/scipy-1.12.0=.,
      -fstack-protector-strong, -fstack-clash-protection, -Wformat, -Werror=format-security,
      -fcf-protection, -Wdate-time, -D_FORTIFY_SOURCE=2
    name: gcc
    version: 13.2.0
  c++:
    args: -g, -O2, -ffile-prefix-map=/build/reproducible-path/scipy-1.12.0=., -fstack-protector-strong,
      -fstack-clash-protection, -Wformat, -Werror=format-security, -fcf-protection,
      -Wdate-time, -D_FORTIFY_SOURCE=2
    commands: c++
    linker: ld.bfd
    linker args: -Wl,-z,relro, -g, -O2, -ffile-prefix-map=/build/reproducible-path/scipy-1.12.0=.,
      -fstack-protector-strong, -fstack-clash-protection, -Wformat, -Werror=format-security,
      -fcf-protection, -Wdate-time, -D_FORTIFY_SOURCE=2
    name: gcc
    version: 13.2.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    args: -fPIC
    commands: gfortran
    linker: ld.bfd
    linker args: -Wl,-z,relro, -fPIC
    name: gcc
    version: 13.2.0
  pythran:
    include directory: ../../../../usr/lib/python3/dist-packages/pythran
    version: 0.16.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /usr/bin/python3.11
  version: '3.11'
```

--- a/scipy/linalg/tests/test_decomp.py
+++ b/scipy/linalg/tests/test_decomp.py
@@ -181,7 +181,8 @@ class TestEig:
         assert_equal(w, np.inf)
         assert_allclose(vr, 1)
 
-    def _check_gen_eig(self, A, B, atol_homog=1e-13, rtol_homog=1e-13):
+    def _check_gen_eig(self, A, B, atol_homog=1e-13, rtol_homog=1e-13,
+                                   atol=1e-13, rtol=1e-13):
         if B is not None:
             A, B = asarray(A), asarray(B)
             B0 = B
@@ -230,7 +231,7 @@ class TestEig:
         for i in range(res.shape[1]):
             if np.all(isfinite(res[:, i])):
                 assert_allclose(res[:, i], 0,
-                                rtol=1e-13, atol=1e-13, err_msg=msg)
+                                rtol=rtol, atol=atol, err_msg=msg)
 
         # try to consistently order eigenvalues, including complex conjugate pairs
         w_fin = w[isfinite(w)]
@@ -269,7 +270,7 @@ class TestEig:
                    [24, 35, 18, 21, 22]])
 
         with np.errstate(all='ignore'):
-            self._check_gen_eig(A, B, atol_homog=5e-13)
+            self._check_gen_eig(A, B, atol_homog=5e-13, atol=5e-13)
 
     def test_falker(self):
         # Test matrices giving some Nan generalized eigenvalues.

